test/unit/transports/azure_test.rb

```diff
@@ -0,0 +1,146 @@
+    Train::Transports::Azure.new(options)
+  end
+  let(:connection) { transport.connection }
+  let(:options) { connection.instance_variable_get(:@options) }
+  let(:cache) { connection.instance_variable_get(:@cache) }
+  let(:credentials) { connection.instance_variable_get(:@credentials) }
+
+  describe 'options' do
+    it 'defaults to env options' do
+      options[:tenant_id].must_equal 'test_tenant_id'
+      options[:client_id].must_equal 'test_client_id'
+      options[:client_secret].must_equal 'test_client_secret'
+      options[:subscription_id].must_equal 'test_subscription_id'
+    end
+
+    it 'test options override' do
+      transport = transport(subscription_id: '102', client_id: '717')
+      options = transport.connection.instance_variable_get(:@options)
+      options[:tenant_id].must_equal 'test_tenant_id'
+      options[:client_id].must_equal '717'
+      options[:client_secret].must_equal 'test_client_secret'
+      options[:subscription_id].must_equal '102'
+    end
+
+    it 'test url parse override' do
+      transport = transport(host: '999')
+      options = transport.connection.instance_variable_get(:@options)
+      options[:tenant_id].must_equal 'test_tenant_id'
+      options[:subscription_id].must_equal '999'
+    end
+  end
+
+  describe 'platform' do
+    it 'returns platform' do
+      plat = connection.platform
+      plat.name.must_equal 'azure'
+      plat.family_hierarchy.must_equal ['cloud', 'api']
+    end
+  end
+
+  describe 'azure_client' do
+    class AzureResource
+      attr_reader :hash
+      def initialize(hash)
+        @hash = hash
+      end
+    end
+    it 'test azure_client with caching' do
+      connection.instance_variable_set(:@credentials, {})
+      client = connection.azure_client(AzureResource)
+      client.is_a?(AzureResource).must_equal true
+      cache[:api_call].count.must_equal 1
+    end
+
+    it 'test azure_client without caching' do
+      connection.instance_variable_set(:@credentials, {})
+      connection.disable_cache(:api_call)
+      client = connection.azure_client(AzureResource)
+      client.is_a?(AzureResource).must_equal true
+      cache[:api_call].count.must_equal 0
+    end
+
+    it 'test azure_client default client' do
+      client = connection.azure_client
+      client.class.must_equal Azure::Resources::Profiles::Latest::Mgmt::Client
+    end
+  end
+
+  describe 'connect' do
+    it 'validate credentials' do
+      connection.connect
+      credentials[:credentials].class.must_equal MsRest::TokenCredentials
+      credentials[:tenant_id].must_equal 'test_tenant_id'
+      credentials[:client_id].must_equal 'test_client_id'
+      credentials[:client_secret].must_equal 'test_client_secret'
+      credentials[:subscription_id].must_equal 'test_subscription_id'
+    end
+  end
+
+  describe 'parse_credentials_file' do
+    let(:cred_file) do
+      require 'tempfile'
+      file = Tempfile.new('cred_file')
+      info = <<-INFO
```

Comment:
Not sure if leading whitespace matters in `.ini`, but you could use a squiggly heredoc. E.g. `<<~INFO`

Reply:
Yeah I tried that, some of the appvoyer stuff for train was using 2.2 and was bombing. So reverted to the old system.